pkg/operator/ceph/cluster/osd/update.go

```diff
@@ -0,0 +1,345 @@
+
+			message := fmt.Sprintf("Processing OSD %d on PVC %q", osdID, nodeOrPVCName)
+			updateConditionFunc(c.cluster.context, c.cluster.clusterInfo.NamespacedName(), cephv1.ConditionProgressing, v1.ConditionTrue, cephv1.ClusterProgressingReason, message)
+		}
+		if err != nil {
+			errs.addError("%v", errors.Wrapf(err, "failed to update OSD %d", osdID))
+			continue
+		}
+
+		updatedDeployments = append(updatedDeployments, updatedDep)
+		listIDs = append(listIDs, strconv.Itoa(osdID))
+	}
+
+	// when waiting on deployments to be updated, only list OSDs we intend to update specifically by ID
+	listFunc := c.cluster.getFuncToListDeploymentsWithIDs(listIDs)
+
+	failures := updateMultipleDeploymentsAndWaitFunc(c.cluster.context.Clientset, updatedDeployments, listFunc)
+	for _, f := range failures {
+		errs.addError("%v", errors.Wrapf(f.Error, "failed to update OSD deployment %q", f.ResourceName))
+	}
+
+	// If there were failures, don't retry them. If it's a transitory k8s/etcd issue, the next
+	// reconcile should succeed. If it's a different issue, it will always error.
+	c.queue.Remove(osdIDs)
+}
+
+// getOSDUpdateInfo returns an update queue of OSDs which need updated and an existence list of OSD
+// Deployments which already exist.
+func (c *Cluster) getOSDUpdateInfo() (*updateQueue, *existenceList, error) {
+	ctx := context.TODO()
+	namespace := c.clusterInfo.Namespace
+
+	listOpts := metav1.ListOptions{
+		// list only rook-ceph-osd Deployments
+		LabelSelector: fmt.Sprintf("%s=%s", k8sutil.AppAttr, AppName),
+	}
+	deps, err := c.context.Clientset.AppsV1().Deployments(namespace).List(ctx, listOpts)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to query existing OSD deployments to see if they need updated")
+	}
+
+	updateQueue := newUpdateQueueWithCapacity(len(deps.Items))
+	existenceList := newExistenceListWithCapacity(len(deps.Items))
+
+	for i, d := range deps.Items {
+		id, err := getOSDID(&deps.Items[i]) // avoid implicit memory aliasing by indexing
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// all OSD deployments should be marked as existing
+		existenceList.Add(id)
+
+		rookMatch, err := k8sutil.RookVersionLabelMatchesCurrent(d.Labels)
+		if err != nil {
+			logger.Warningf("failed to determine if OSD Deployment %q was created by the current Rook version; assuming it needs updated. %v", d.Name, err)
+			rookMatch = false
+		}
+		cephMatch := false
+		desiredCephImage := c.spec.CephVersion.Image
+		if len(d.Spec.Template.Spec.Containers) == 0 {
+			logger.Warningf("failed to determine if OSD Deployment %q is running the desired Ceph image (%s); assuming it needs updated. deployment has no containers specified", d.Name, desiredCephImage)
+		} else {
+			cephMatch = (d.Spec.Template.Spec.Containers[0].Image == desiredCephImage)
+		}
+
+		// undocumented option to aid debugging
+		if os.Getenv("ROOK_FORCE_OSD_UPDATES") == "yes-i-really-mean-it" {
+			rookMatch = false
+			cephMatch = false
+		}
```

Comment:
This forces OSD deployments to update regardless of whether the rook/ceph version have changed. Is this useful to keep in for testing or if a user needs to force updates?

Reply:
This forces the code to update deployments even if they are already at the updated versions. I think that is separate from forcing updates for unhealthy daemons.

Reply:
I changed the name of the env var to be more descriptive.